Document Google Cloud CLI access

ptgott opened the pull request:
Closes #17584

- Add a guide to configuring and using Google Cloud CLI access via the Teleport Application Service.

- Since we now have three guides to setting up secure access to cloud APIs, create a new subsection of the "Application Access" docs section related to cloud provider APIs.

ptgott committed:
Document Google Cloud CLI access

Closes #17584

- Add a guide to configuring and using Google Cloud CLI access via the
  Teleport Application Service.

- Since we now have three guides to setting up secure access to cloud
  APIs, create a new subsection of the "Application Access" docs section
  related to cloud provider APIs.

zmb3 commented on docs/config.json:
```diff
@@ -595,7 +595,25 @@
         },
         {
-          "title": "Access Guides",
+          "title": "Securing Cloud APIs",
```
```suggestion
          "title": "Securing Cloud Consoles",
```

What do you think about saying "consoles" instead? To me, it's more obvious that this is about access to the major hyperscalers and not any arbitrary cloud API.

ptgott replied:
@zmb3 Would "Securing Cloud Provider APIs" be clearer?

benarent replied:
I personally think consoles is better. Azure calls it a Portal but it's clear that what a console is. API feels a little limiting.

ptgott replied:
The only console we can protect with the Application Service is the AWS console. For Azure and Google Cloud, the Application Service protects CLI tools by proxying requests to the cloud provider's API. I think "Securing Cloud Provider APIs" may be the most precise description.

zmb3 commented on docs/pages/application-access/introduction.mdx:
```diff
@@ -37,53 +36,45 @@ internal dashboards and applications, such as:
 - [Interactive Lab](https://play.instruqt.com/teleport/invite/rgvuva4gzkon): Try Teleport using our guided Teleport Application Access lab.
 
-## JWT guides
+## Use Teleport-signed JSON web tokens
```
Should JSON Web Tokens be capitalized? It's a proper noun, isn't it? jwt.io seems to capitalize it

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+```
+
+This command uses the `--set-gcp-service-accounts` flag to add Google Cloud
+service accounts to a user. You can assign `--set-gcp-service-accounts` to
+multiple service account URIs, separated by commas.
````
```suggestion
service accounts to a user. You can use `--set-gcp-service-accounts` to
assign multiple service account URIs, separated by commas.
```

ptgott replied:
I've used a variation of this that makes it clearer that the comma-separated list is a value of `--set-gcp-service-accounts`

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+    app_labels:
+      '*': '*'
+    gcp_service_accounts:
+      - teleport-vm-lister@my-project.iam.gserviceaccount.com
```
```suggestion
      - teleport-vm-lister@${PROJECT_ID?}.iam.gserviceaccount.com
```

ptgott replied:
Since this is part of a YAML snippet, I'm not sure a shell variable would make sense here. Should I make `my-project` more prominent as a placeholder value by capitalizing it?

Tener replied:
Ah, I thought YAML snippets are treated the same way as the shell snippets. I guess capitalization works, too, anything really to make sure this gets updated.

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+
+```text
+Logged into GCP app "google-cloud-cli".
+Your service account: teleport-google-cloud-cli@my-project.iam.gserviceaccount.com
````
```suggestion
Your service account: teleport-vm-lister@${PROJECT_ID?}.iam.gserviceaccount.com
```

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+
+### Use Google Cloud CLI applications without `tsh`
+
+In addition to running `gcloud` commands via `tsh`, you can grant secure access
```
```suggestion
In addition to running `gcloud` and `gsutil` commands via `tsh`, you can grant secure access
```

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+To start the local proxy, run the following `tsh` command:
+
+```code
+$ tsh proxy gcloud 
````
I wonder if we should rename the proxy command to `tsh proxy gcp` instead, as GCP is the cloud and `gcloud` is one of the command line tools. Thoughts?

ptgott replied:
Good question—it looks like Google has been moving away from the "GCP" usage in favor of "Google Cloud" (e.g., on the homepage: https://cloud.google.com/), but it seems like "GCP" is still well known and crops up every now and then in their docs (e.g., the diagram here: https://cloud.google.com/docs/overview#global_regional_and_zonal_resources).

I think `tsh proxy gcp` makes sense.

Tener replied:
Interesting, I didn't realize that Google seems to be moving away from the "GCP" term. I guess that makes `tsh proxy gcloud` a good name after all. I'll think about adding `tsh proxy gcp` as an alias.

ptgott replied:
Weirdly, I can't find any blog post or similar announcement that there's a formal name change, just pages that used to say "Google Cloud Platform" that now say "Google Cloud"

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+for Google Cloud's APIs:
+
+```text
+Started GCP proxy on http://127.0.0.1:50614.
+To avoid port randomization, you can choose the listening port using the --port flag.
+
+Use the following credentials and HTTPS proxy setting to connect to the proxy:
+
+  export CLOUDSDK_AUTH_ACCESS_TOKEN=00000000000000000000000000000000
+  export CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE=/Users/myuser/.tsh/keys/teleport.example.com/myuser-app/teleport.example.com/google-cloud-cli-localca.pem
+  export HTTPS_PROXY=http://127.0.0.1:50614
````
There are more env vars now.

```suggestion
Started GCP proxy on http://127.0.0.1:50614.
To avoid port randomization, you can choose the listening port using the --port flag.

Use the following credentials and HTTPS proxy setting to connect to the proxy:

  export BOTO_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/google-cloud-cli/00000000_boto.cfg
  export CLOUDSDK_AUTH_ACCESS_TOKEN=00000000000000000000000000000000
  export CLOUDSDK_CONFIG=/Users/myuser/.tsh/gcp/teleport.example.com/google-cloud-cli/gcloud
  export CLOUDSDK_CORE_CUSTOM_CA_CERTS_FILE=/Users/myuser/.tsh/keys/teleport.example.com/myuser-google-cloud-cli/teleport.example.com/google-cloud-cli-localca.pem
  export CLOUDSDK_CORE_PROJECT=${PROJECT_ID?}
  export HTTPS_PROXY=http://127.0.0.1:50614
```

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+```code
+$ gcloud compute instances list
+```
+
````
`gsutil ls` would work in this shell now too, but perhaps this is redundant to point out here.

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+
+While you could run `gcloud iam service-accounts create` in a shell earlier in this
+guide, after entering the `export` commands printed by `tsh proxy gcloud`, this
+command results in authorization issues:
```
```suggestion
command will be run as a restricted user, which results in authorization issues:
```

ptgott replied:
👍 I've also broken up this sentence a bit to make room for the new text

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+
+```code
+$ gcloud iam service-accounts create demo-service-account
+ERROR: (gcloud.iam.service-accounts.create) User [myuser] does not have permission to access projects instance [myproject] (or it may not exist): Permission 'iam.serviceAccounts.create' denied on resource (or it may not exist).
````
Nit: update `[myuser]` and `[myproject]`?

ptgott replied:
What would you recommend here? Since I was using `myuser` and `myproject` throughout the page, I think leaving these placeholders would make sense

Tener replied:
Ah, I thought that you could use variables here, so the user sees their own user/project here. If that isn't possible then I think we can leave it as it is.

ptgott replied:
Sounds good—right now it looks like the `Var` component only renders within lines beginning with `$` in `code` snippets.

Tener commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+  Requests](../../access-controls/access-requests/role-requests.mdx) and [Access
+  Request plugins](../../access-controls/access-request-plugins/index.mdx).
+
+- You can proxy any `gcloud` command via Teleport. For a full reference of
```
I'd add `gsutil` reference in this or a sibling point.

Tener commented on docs/pages/application-access/introduction.mdx:
```diff
@@ -1,13 +1,12 @@
-internal dashboards and applications, such as:
+cloud provider APIs and internal applications. Examples include: 
 
+- The AWS management console.
```
Should we add another bullet point for Azure and GCP cloud APIs?

ptgott replied:
I've added a bullet for CLIt tools for all three major cloud providers

smallinsky commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+---
+title: "Protect Google Cloud CLIs with Teleport"
+description: How to enable secure access to Google Cloud CLIs.
```
In fact the Google Cloud Platform is accessed via google CLI 

so maybe instead
> Protect Google Cloud CLIs with Teleport

the title should sound something like: "Protect Google Cloud Platform Access with Teleport"

Tener reviewed (commented):
A few rather minor points.

smallinsky commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+within `app_service.apps` is an application configuration. 
+
+In the example above, we have enabled Google Cloud CLI access by registering an
+application called `google-cloud-cli` with the `cloud` field set to to `GCP`.
```
```suggestion
application called `google-cloud-cli` with the `cloud` field set to `GCP`.
```

ptgott committed:
Respond to PR feedback

- Capitalize "JSON Web Tokens"
- Edit the AWS guide's nav bar entry
- Fix garbled section in the App Access intro page
- Improve the service account name we use in the how-to guide
- Clarify the way the App Service uses a service account
- Add new env vars to the `tsh proxy` command output
- Other small changes in response to suggestions/comments

ptgott commented on docs/config.json:
```diff
@@ -595,7 +595,25 @@
+              "slug": "/application-access/cloud-apis/google-cloud/"
+            },
+            {
+              "title": "AWS Access",
```
Sounds good to me!

ptgott commented:
## To do
- [ ] Document `tsh gsutil` in the how-to guide for GCP CLI access as well as in any relevant reference docs

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+<Admonition type="tip" title="Using existing service accounts">
+
+In this guide, we will demonstrate Google Cloud CLI access by creating a service
+account for Teleport users to authenticate as, `teleport-vm-lister`. If you would
```
That works for me!

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+
+### Create a service account for the Application Service
+
+The Application Service uses a Google Cloud service account to generate signed
```
I've suggested another wording that hopefully makes this paragraph clearer. I think the original version was too compact.

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+
+## Step 2/4. Deploy the Teleport Application Service
+
+At this point, you have created a service account for the Teleport Application
+Service and enabled this service account to impersonate the service accounts you
+would like Teleport users to access. 
+
+In this step, you will attach the service account you created for the
+Application Service to your Google Compute Engine VM, then run the Teleport
+Application Service.
```
That makes sense! I'll make time to implement this

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+Now that you have created a Google Cloud service account and attached a role to
+it, associate your service account with the Teleport Application Service VM.
+
+Stop your VM so you can attach your service account to it:
```
This sounds good—adding to the to-do list

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+
+ERROR: exit status 1
+```
+
````
👍 I've added that to my list of things to do later this week

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,615 @@
+To do this, use `tsh` to start a local proxy that forwards traffic from your CLI
+application to the Teleport Application Service. The Application Service uses
+the `teleport-google-cloud-cli` service account we created earlier to fetch an
+authentication token from Google Cloud. Your CLI application uses this token to
+authenticate requests to Google Cloud's APIs.
```
I've added this information to the first paragraph of the "Create a service account for the Application Service" section, since this is where we first mention the token after the intro section.

tigrato commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
````diff
@@ -0,0 +1,615 @@
+
+```yaml
+kind: role
+version: v5
````
Yes, we should upgrade our docs to `v6` to keep consistency since Kubernetes RBAC will require `v6`

ptgott committed:
Edit instructions for starting the VM

Add instructions to launch a new VM with a service account attached.
This addresses feedback on the PR.

ptgott committed:
Use more precise terminology for service accounts

This addresses feedback on the PR

ptgott committed:
Add information about `tsh gsutil`

This adds to both the Google Cloud API access how-to guide and the CLI
reference.

ptgott commented:
@Tener I think I've addressed your feedback. I mentioned `tsh gsutil` in the how-to guide, but in asides rather than within the core step-by-step flow. The reason is that, while a user following the instructions would definitely have a VM to list (the one that runs the App Service), they wouldn't necessarily have a Google Cloud Storage bucket.

@alexfornuto This should be ready for copy review when you have a moment, thanks!

Tener reviewed (approved):
🚢

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -244,6 +244,70 @@ List all available applications:
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+| - | - | - | - |
+| `--app` | Currently logged in Google Cloud application | The name of a Google Cloud application as listed via `tsh apps ls`. | The Google Cloud application to run the command against, if logged in to multiple Google Cloud applications. |
```
Employs `<br/>` tags to prevent `--app` from being broken across lines, until https://github.com/gravitational/docs/issues/126 is resolved.

```suggestion
| `--app` | Currently logged in<br/>Google Cloud application | The name of a Google Cloud<br/>application as listed via `tsh apps ls`. | The Google Cloud application to run<br/>the command against, if logged in<br/> to multiple Google Cloud applications. |
```

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -244,6 +244,70 @@ List all available applications:
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+| - | - | - | - |
+| `--app` | Currently logged in Google Cloud application | The name of a Google Cloud application as listed via `tsh apps ls`. | The Google Cloud application to run the command against, if logged in to multiple Google Cloud applications. |
```
Same as above:

```suggestion
| `--app` | Currently logged in<br/>Google Cloud application | The name of a Google Cloud<br/> application as listed via `tsh apps ls`. | The Google Cloud application to run<br/>the command against, if logged in to multiple Google Cloud applications. |
```

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -638,6 +702,40 @@ $ tsh proxy aws --port=10700 --app=awsapp2
+
+#### Flags
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
```
This table needs formatting as well, but I wasn't able to find the magic `<br/>` placement to fix it.

alexfornuto commented on docs/pages/reference/cli.mdx:
```diff
@@ -638,6 +702,40 @@ $ tsh proxy aws --port=10700 --app=awsapp2
+| - | - | - | - |
+| `--app` | Currently logged in Google Cloud app | string | Optional name of the Google Cloud application (as shown in `tsh apps ls`) to use if logged in to multiple |
+| `-p, --port` | none | port number | Specify the source port for the local proxy |
+| ` -e, --endpoint-url` | HTTP Proxy | Endpoint URL | Run the local proxy to serve as an AWS endpoint URL. If not specified, the local proxy serves as an HTTPS proxy. |
```
I'm guessing:

```suggestion
| ` -e, --endpoint-url` | HTTP Proxy | Endpoint URL | Run the local proxy to serve as a Google Cloud endpoint URL. If not specified, the local proxy serves as an HTTPS proxy. |
```

alexfornuto commented on docs/pages/application-access/cloud-apis.mdx:
```diff
@@ -0,0 +1,20 @@
+---
+title: "Securing Access to Cloud APIs"
+description: "How to use Teleport to achieve secure access to managing cloud providers"
```
OR:

```suggestion
description: "How to use Teleport to achieve secure access to manage cloud providers"
```

alexfornuto replied:
P.P.S: Is the convention to include periods in descriptions? I feel like I've seen both.

ptgott replied:
👍

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,689 @@
+these requests using tokens retrieved from Google Cloud. This enables Teleport
+operators to control the service accounts that users can assume in order to
+interact with Google Cloud APIs.
+
```
One day soon!

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,689 @@
+
+- Either a Google Compute Engine VM where you will run the Teleport Application
+  Service *or* permissions to create VMs in your Google Cloud project. If you
+  are using a pre-existing VM, it must be running a Linux distribution. You must
+  have permissions to attach service accounts to Google Compute Engine VMs.
```
Rephrased

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,689 @@
+
+<Admonition type="tip" title="Using existing service accounts">
+
+In this guide, we will demonstrate Google Cloud CLI access by creating a service
+account for Teleport users to authenticate as, `teleport-vm-viewer`. If you would
+like to enable access to any existing service accounts in your Google Cloud
+project, you can replace `teleport-vm-viewer` with these as you work through the
+guide.
```
Other code snippets assume that you have set up this service account to view VMs, so I'm not sure we want to make it _to_ easy to use a custom service account name.

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,689 @@
+
+### Install the Teleport Application Service
+
+Run the following commands on the host where you will install the Teleport
+Application Service:
```
Made a suggestion

ptgott commented on docs/pages/application-access/cloud-apis/google-cloud.mdx:
```diff
@@ -0,0 +1,689 @@
+### Configure the Teleport Application Service
+
+On the host where you will run the Teleport Application Service, create a file
+at `/etc/teleport.yaml` with the following content:
```
It's a good idea, but I'm not sure we need it in this case—I think it would make sense to do this in a separate PR for the sake of time.

ptgott committed:
Respond to PR feedback

ptgott commented:
@alexfornuto I'm blocking on your feedback since you provided the last review, but please let me know if this is good to merge without it. Thanks!

Commit:
use nobr tags

Commit:
fix multiline command

alexfornuto reviewed (approved):
Made a couple more edits, now it's 👍 to :shipit:
